backend/onyx/tools/tool_implementations/document/document_editor_tool.py

```
@@ -215,16 +251,56 @@ def _run(
 
             # When using structured_response_format, the content is already a dict
             # Ensure the required fields are present
-            if "edited_text" not in edit_result or "summary" not in edit_result:
-                raise ValueError("LLM response missing required fields: edited_text and/or summary")
+            if "changes" not in edit_result or "summary" not in edit_result:
+                raise ValueError(
+                    "LLM response missing required fields: changes and/or summary"
+                )
+
+            # Apply the changes to the original text
+            edited_text = self.document_content
+            changes = edit_result.get("changes", [])
+
+            # Sort changes in reverse order to avoid position shifting issues
+            changes.sort(key=lambda x: len(x.get("context_before", "")), reverse=True)
+
+            for change in changes:
+                context_before = change.get("context_before", "")
+                context_after = change.get("context_after", "")
+
+                # Find the position to make the change
+                start_pos = edited_text.find(context_before)
+                if start_pos == -1:
+                    continue
+
+                start_pos += len(context_before)
+                end_pos = edited_text.find(context_after, start_pos)
+                if end_pos == -1:
+                    continue
```

not sure that it should silently fail, ideally this would cause the entire set of changes to be rejected, and tool result message would indicate the failure back to the model to try again.